### DISCUSSION TITLE
upstream hijacking

### DISCUSSION THREAD

**Discussion by pschrammel** [puma/puma#3100] [General]
**Is your feature request related to a problem? Please describe.**
Uploading big files (> 100 GB) is a pain as it's first buffered on the filesystem.

**Describe the solution you'd like**
On having received the headers the stream could be passed and routed to the application.  The application then could handle each block of the body.

**Describe alternatives you've considered**
I tried nodejs as they have a proper streaming api/io but it's a pain. I used goliath server in the past but it's unmaintained.

**Additional context**
The rough context is a fileserver (webdav, ....). Where ppl can upload files but the real storage is handled by s3. The parsing of the body or pass through should be somehow selective  (some paths' body  should be parsed, some not ....).

**Comment by dentarg** [comment]
Isn't this possible? At least since https://github.com/puma/puma/pull/2740

**Comment by dentarg** [comment]
See https://github.com/rack/rack/blob/main/UPGRADE-GUIDE.md#response-bodies-can-be-used-for-bi-directional-streaming

**Comment by MSP-Greg** [comment]
I think the question is about a way to stream the request body, not the response body.  I haven't had time to think about it.

It is somewhat similar to a partial hijack, in that we'd send the env to the app with the client socket included.  Might get a bit tricky to have the app then 'return' the socket along with the app 'reply', containing the response. to the request...

**Reply by dentarg** [reply]
Yeah, I realised... but https://github.com/rack/rack/blob/main/UPGRADE-GUIDE.md#response-bodies-can-be-used-for-bi-directional-streaming also reads from `stream`, what's that if not the request?

**Reply by MSP-Greg** [reply]
I was too quick to reply.  Partial hijack is not similar to this.

1. Both full and partial hijack happen after `app.call`.  Full hijack happens right after, partial happens after Puma has sent the response headers to the client.  So, full hijack might work, but I'm not sure about stopping the request reading/parsing at the headers/body boundary.

2. AFAIK, the only 'bi-directional-streaming' that exists is with WebSocket, and once it's in place, there's no way to return to 'http' communication.  Any request that is uploading a large amount of data (via http) is expecting an http response?

**Reply by dentarg** [reply]
I see

I'll continue to link to the Rack 3 upgrade guide... sounds like streaming the request body will need Rack >=3?, from 
 https://github.com/rack/rack/blob/8483b1034de7c9b18201f60fe8735fc83d32ba21/UPGRADE-GUIDE.md#rackinput-is-no-longer-required-to-be-rewindable

> Previously, `rack.input` was required to be rewindable, i.e. `io.seek(0)` but this was only generally possible with a file based backing, which prevented efficient streaming of request bodies. Now, `rack.input` is not required to be rewindable.

**Comment by nateberkopec** [comment]
Do you have a reverse proxy (i.e. Nginx) in front of Puma, at least? Might not solve the problem but it will probably help.

**Reply by pschrammel** [reply]
yes, but I explicitly don't want to have any kind of buffering (imagine implementing s3 server in ruby ....okay, sounds stupid). If somebody could outline where to look at / what / how to hack I'd give it a try and do some experiments.

**Comment by nateberkopec** [comment]
Moved to discussion as next action is not clear.

**Reply by nateberkopec** [reply]
Man, I really don't like the formatting for Q&A...

**Reply by nateberkopec** [reply]
The threading-always-by-default is pretty weird.
